crates/ln-dlc-node/src/ldk_node_wallet.rs

```diff
@@ -182,55 +191,118 @@ where
+        });
+
+        // Get temporarily reserved UTXOs from in-memory storage.
         let mut reserved_outpoints = self.locked_outpoints.lock();
 
-        // filter reserved utxos from all known utxos to not accidentally double spend and those who
-        // have actually been spent already
+        let utxos = self.get_utxos()?;
+
+        // Filter out reserved and spent UTXOs to prevent double-spending attempts.
         let utxos = utxos
             .iter()
             .filter(|utxo| !reserved_outpoints.contains(&utxo.outpoint))
             .filter(|utxo| !utxo.is_spent)
             .collect::<Vec<_>>();
 
-        let mut utxos = utxos
-            .into_iter()
-            .map(|x| UtxoWrap {
-                utxo: Utxo {
-                    tx_out: x.txout.clone(),
-                    outpoint: x.outpoint,
-                    address: Address::from_script(
-                        &x.txout.script_pubkey,
-                        self.bdk_lock().network(),
-                    )
-                    .expect("to be a valid address"),
-                    redeem_script: Default::default(),
-                    reserved: false,
-                },
+        let candidates = utxos
+            .iter()
+            .map(|utxo| {
+                let tx_in = TxIn {
+                    previous_output: utxo.outpoint,
+                    ..Default::default()
+                };
+
+                // Inspired by `rust-bitcoin:0.30.2`.
+                let segwit_weight = {
+                    let legacy_weight = {
+                        let script_sig_size = tx_in.script_sig.len();
+                        (36 + VarInt(script_sig_size as u64).len() + script_sig_size + 4) * 4
+                    };
+
+                    legacy_weight + tx_in.witness.serialized_len()
+                };
+
+                // The 10101 wallet always generates SegWit addresses.
+                //
+                // TODO: Rework this once we used Taproot.
+                let is_witness_program = true;
+
+                Candidate::new(utxo.txout.value, segwit_weight as u32, is_witness_program)
             })
             .collect::<Vec<_>>();
 
-        // select enough utxos for our needs
-        let selected_local_utxos = select_coins(amount, 20, &mut utxos).with_context(|| {
-            format!("Could not reach target of {amount} sats with given UTXO pool")
-        })?;
-
-        // update our temporarily reserved utxos with the selected once.
-        // note: this storage is only cleared up on a restart, meaning, if the protocol
-        // fails later on, the utxos will remain reserved
-        if lock_utxos {
-            for utxo in selected_local_utxos.clone() {
-                reserved_outpoints.push(utxo.utxo.outpoint);
+        // This is a standard base weight (without inputs or change outputs) for on-chain DLCs. We
+        // assume that this value is still correct for DLC channels.
+        let funding_tx_base_weight = 212;
```

Comment:
The weight depends on how many inputs will be selected, so very likely this will be wrong if there are more than 1 input per party.

Reply:
True, although this is the _base_ weight. According to `bdk_coin_select`:

```rust
/// The `base_weight` is the weight of the transaction without any inputs and without a change
/// output.
```

Reply:
oh, thanks for the explanation.